### ISSUE TITLE
Project system data flow 'Workspace update handler 17344132' closed because of an exception: This program location is thought to be unreachable. 

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#70134]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Project-system-data-flow-Workspace-upda/10474895)._

---
VS goldbar: "Visual Studio ran into an unexpected problem with one or more projects. ..."

Log contents:
```
=====================
9/25/2023 5:51:26 PM
LimitedFunctionality
System.AggregateException: Project system data flow 'Workspace update handler 17344132' closed because of an exception: This program location is thought to be unreachable. File='/_/src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs' Line=76.
---> (Inner Exception #0) System.InvalidOperationException: This program location is thought to be unreachable. File='/_/src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs' Line=76
   at Microsoft.CodeAnalysis.SolutionState.WithDocumentContentsFrom(DocumentId documentId, DocumentState documentState)
   at Microsoft.CodeAnalysis.Solution.WithDocumentContentsFrom(DocumentId documentId, DocumentState documentState)
   at Microsoft.CodeAnalysis.Workspace.<SetCurrentSolutionAsync>g__UpdateExistingDocumentsToChangedDocumentContents|27_2(Solution solution, DocumentId changedDocumentId, HashSet`1 processedDocuments)
   at Microsoft.CodeAnalysis.Workspace.<SetCurrentSolutionAsync>g__UnifyLinkedDocumentContents|27_0(Solution oldSolution, Solution newSolution)
   at Microsoft.CodeAnalysis.Workspace.<>c.<SetCurrentSolutionAsync>b__27_3(Solution oldSolution, ValueTuple`5 data)
   at Microsoft.CodeAnalysis.Workspace.<SetCurrentSolutionAsync>d__29`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Workspace.<SetCurrentSolutionAsync>d__27.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Workspaces.ProjectSystem.ProjectSystemProjectFactory.<ApplyBatchChangeToWorkspaceMaybeAsync>d__46.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Workspaces.ProjectSystem.ProjectSystemProject.<OnBatchScopeDisposedMaybeAsync>d__102.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Workspaces.ProjectSystem.ProjectSystemProject.BatchScope.<DisposeAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.LanguageServices.Workspace.<>c__DisplayClass43_1`1.<<OnProjectChangedAsync>g__ApplyInBatchAsync|1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Threading.ReentrantSemaphore.StackSemaphore.<>c__DisplayClass3_0.<<ExecuteAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Threading.ReentrantSemaphore.StackSemaphore.<ExecuteAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.OnceInitializedOnceDisposedUnderLockAsync.<ExecuteUnderLockCoreAsync>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.LanguageServices.Workspace.<OnBuildUpdateAsync>d__41.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.LanguageServices.Workspace.<>c__DisplayClass39_0.<<OnWorkspaceUpdateAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.LanguageServices.Workspace.<>c__DisplayClass39_0.<<OnWorkspaceUpdateAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ProjectSystem.LanguageServices.Workspace.<OnWorkspaceUpdateAsync>d__39.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.DataReceivingBlockSlim`1.<ProcessInputQueueAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.CommonProjectSystemTools.Rethrow(Exception ex)
   at Microsoft.VisualStudio.ProjectSystem.ExceptionFilter.<>c__DisplayClass2_0.<Guard>g__Action|0()
   at GuardMethodClass.GuardMethod(Func`1 , Func`2 , Func`2 )
<--- (Inner Exception #0) 
===================
```

See attached traces.

---
#### Black Elijah (ST-IN ENG3-TW) on 9/25/2023, 09:30 PM: 

(private comment, text removed)
#### Black Elijah (ST-IN ENG3-TW) on 9/25/2023, 09:33 PM: 

(private comment, text removed)
#### Feedback Bot on 9/25/2023, 09:56 PM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi Ah, I see the problem: all types of documents contribute to our file name -> ID map, including additional files and .editorconfig files. For the most part those should be disjoint in practice, but if somebody makes a mistake like this bug we'll end up with a source file which has related IDs that aren't source files but a different type of document.

**Comment by jasonmalinowski** [comment]
(fixing this now)

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski  i wonder if we'll see this being a similar issue for the other cases.  if we can collect dumps, that will probably help a lot.

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi We do have dump collection for this issue but unfortunately we haven't gotten any hits with actual dumps yet.

This might also be an indirect cause of some of other uptick of failures we're seeing in the project system area: the project system code maintains some other state that it expects to be in sync with the current solution. If the project system is applying a batch and then SetCurrentSolution _fails_, that state it has is now out of sync with CurrentSolution, and at that point bets are somewhat off.

This unfortunately still doesn't explain the syntax-trees-out-of-sync-with-compilations bug though....

**Comment by Eli-Black-Work** [comment]
Let me now if you need additional dumps or anything 🙂

**Comment by Eli-Black-Work** [comment]
Wow, I was not expecting the submission of this bug to kick off such a firestorm 😄

**Comment by Eli-Black-Work** [comment]
BTW, the reason that I discovered this crash was because of how I created the `.editorconfig` file: In VS, I copied-pasted an existing `.cs` file and renamed it to `.editorconfig`. This caused VS to modify the `.csproj` and add `<Compile>` for `.editorconfig`

I guess that behaviour makes sense, but it surprised me. I would have expected VS to automatically compile all `.cs` files, and if I copy-paste a `.cs` file and change its extension, I'd expect it to automatically not participate in compilation.

**Comment by jasonmalinowski** [comment]
> Wow, I was not expecting the submission of this bug to kick off such a firestorm 😄

We've had a few bugs that represent potentially corrupted state in our core data structures, specifically #67442 that we had been trying to track down and we had suspected that whatever might be causing this issue was causing that too, so we were investigating both. We've seen this failure in telemetry (where all we get is a callstack), and for unclear reasons our dump collection mechanisms weren't giving us actual dumps. So luckily you having a repro step made this quick to diagnose, and then once I realized it was a regression back in February or so that also means we can potentially rule out some other code that went in at the same time that we thought was at fault.

In short, your repro saved us a lot of time. Should you ever find yourself on the Microsoft campus in Redmond, stop by and we can thank you personally. :smile:

**Comment by Eli-Black-Work** [comment]
> So luckily you having a repro step made this quick to diagnose, and then once I realized it was a regression back in February or so that also means we can potentially rule out some other code that went in at the same time that we thought was at fault.
> 
> In short, your repro saved us a lot of time. Should you ever find yourself on the Microsoft campus in Redmond, stop by and we can thank you personally. 😄

Oh, very cool! ^_^  And ty, I'll keep that in mind 😁

**Comment by Eli-Black-Work** [comment]
> > BTW, the reason that I discovered this crash was because of how I created the .editorconfig file: In VS, I copied-pasted an existing .cs file and renamed it to .editorconfig. This caused VS to modify the .csproj and add  for .editorconfig
> 
> Pinging @tmeschter and @lifengl whether this behavior is intended or not.

@tmeschter and @lifengl, I'd still be interested in knowing if this behaviour is intentional 🙂  I think this might be a regression.

**Comment by tmeschter** [comment]
Possibly what happens is:

1. You copy the .cs file.
2. The project system sees this as a new `<Compile>` item due to the standard globs for .cs files.
3. You rename the item to .editorconfig.
4. The project system notices that the `<Compile>` item is no longer covered by the globa for .cs files (since it no longer ends in .cs).
5. It "helpfully" adds an explicit `<Compile>` item to the project file to preserve the item type.

In which case this is a case of a series of more-or-less reasonable steps leading to an absurd outcome. I can't really say this is _intentional_, but it isn't really a bug either.

**Comment by tmeschter** [comment]
Thinking about this a little more, the behavior makes more sense when you have an explicit, individual `<Compile>` item in your project file. In that case the item type (Compile) is very clear and separate from the file path/name.

It makes less sense when the item comes from a standard glob in the SDK (as it does here). In that case the item type *isn't* separate from the file path/name; rather the name determines the type. If we can detect that case it probably makes more sense to make no modifications to the project file as part of the rename.

**Comment by Eli-Black-Work** [comment]
@tmeschter Thanks! Those are my thoughts as well 🙂

I believe in my case, I didn't have any `<Compile>` items until I'd renamed the file to `.editorconfig` (at which point the `<Compile>` was automatically added).

Should I file a new issue to track having `<Compile>` not be automatically added for files that didn't already have it?

**Comment by jasonmalinowski** [comment]
@Eli-Black-Work Filing something seems reasonable; @tmeschter is this something that could be filed in the project system repo or really needs to go to core CPS so file a VS feedback ticket?
